core/include/vecmem/memory/instrumenting_memory_resource.hpp

```diff
@@ -0,0 +1,155 @@
+         * @param[in] s The size of the request.
+         * @param[in] a The alignment of the request.
+         * @param[in] p The pointer that was returned or deallocated.
+         * @param[in] ns The time taken to perform the request in nanoseconds.
+         */
+        memory_event(type t, std::size_t s, std::size_t a, void* p,
+                     std::size_t ns)
+            : m_type(t), m_size(s), m_align(a), m_ptr(p), m_time(ns) {}
+
+        type m_type;
+
+        std::size_t m_size;
+        std::size_t m_align;
+
+        void* m_ptr;
+
+        std::size_t m_time;
+    };
+
+    /**
+     * @brief Constructs the instrumenting memory resource.
+     *
+     * @param[in] upstream The upstream memory resource to use.
+     */
+    instrumenting_memory_resource(memory_resource& upstream);
+
+    /**
+     * @brief Get the total size of outstanding requests.
+     */
+    std::size_t get_outstanding(void) const;
+
+    /**
+     * @brief Return a list of memory allocation and deallocation events in
+     * chronological order.
+     */
+    const std::vector<memory_event>& get_events(void) const;
+
+    /**
+     * @brief Add a pre-allocation hook.
+     *
+     * Whenever memory is allocated, all pre-allocation hooks are exectuted.
+     * This happens before we know whether the allocation was a success or not.
+     *
+     * The function passed to this function should accept the size of the
+     * request as the first argument, and the alignment as the second.
+     */
+    void add_pre_allocate_hook(std::function<void(std::size_t, std::size_t)> f);
+
+    /**
+     * @brief Add a post-allocation hook.
+     *
+     * Whenever memory is allocated, all post-allocation hooks are exectuted.
+     * This happens after we know whether the allocation was a success or not,
+     * and the pointer that was returned.
+     *
+     * The function passed to this function should accept the size of the
+     * request as the first argument, the alignment as the second, and the
+     * pointer to the allocated memory as the third argument.
+     */
+    void add_post_allocate_hook(
+        std::function<void(std::size_t, std::size_t, void*)> f);
+
+    /**
+     * @brief Add a pre-deallocation hook.
+     *
+     * Whenever memory is deallocated, all pre-deallocation hooks are
+     * exectuted.]
+     *
+     * The function passed to this function should accept the pointer to
+     * allocate as its first argument, the size of the request as the second
+     * argument, and the alignment as the third.
+     */
+    void add_pre_deallocate_hook(
+        std::function<void(void*, std::size_t, std::size_t)> f);
+
+private:
+    virtual void* do_allocate(std::size_t, std::size_t) override;
+
+    virtual void do_deallocate(void* p, std::size_t, std::size_t) override;
+
+    virtual bool do_is_equal(const memory_resource&) const noexcept override;
+
+    memory_resource& m_upstream;
```

Comment:
You documented all the other member variables. Why not this one? :stuck_out_tongue: